plugin/gcp/pubsubTopic.go

```diff
@@ -0,0 +1,173 @@
+	}
+
+	return plugin.ToJSONObject(pubsubTopicMetadata{
+		pubsubTopicAttrMeta: pubsubTopicAttrMeta{PublishSettings: t.topic.PublishSettings},
+		TopicConfig:         cfg,
+		Subscriptions:       subs,
+	}), nil
+}
+
+func (t *pubsubTopic) Delete(ctx context.Context) (bool, error) {
+	return true, t.topic.Delete(ctx)
+}
+
+// A ReadCloser that subscribes to a topic and buffers all messages that appear there.
+type pubsubTopicWatcher struct {
+	*pubsub.Subscription
+	sync.Mutex
+	*list.List
+	context.Context
+	context.CancelFunc
+	mux sync.Mutex
+	err error
+}
+
+func (t *pubsubTopic) newPubsubTopicWatcher(ctx context.Context) (*pubsubTopicWatcher, error) {
+	sub, err := t.client.CreateSubscription(ctx, "wash-temp-subscribe-"+t.topic.ID(), pubsub.SubscriptionConfig{
+		Topic:            t.topic,
+		AckDeadline:      10 * time.Second,
+		ExpirationPolicy: 24 * time.Hour,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	subCtx, cancel := context.WithCancel(context.Background())
+	watcher := &pubsubTopicWatcher{Subscription: sub, List: list.New(), Context: subCtx, CancelFunc: cancel}
+
+	bufferMessages := func(_ context.Context, msg *pubsub.Message) {
+		msg.Ack()
+		watcher.Lock()
+		watcher.PushBack(msg)
+		watcher.Unlock()
+	}
+	go func() {
+		watcher.mux.Lock()
+		watcher.err = sub.Receive(subCtx, bufferMessages)
+		watcher.mux.Unlock()
+	}()
+	return watcher, nil
+}
+
+func (w *pubsubTopicWatcher) Read(p []byte) (int, error) {
+	// If there are outstanding messages, return one.
+	// If not, check if the context is done before returning.
+	var msg *pubsub.Message
+	w.Lock()
+	if w.Len() > 0 {
+		e := w.Front()
+		w.Remove(e)
+		msg = e.Value.(*pubsub.Message)
+	}
+	defer w.Unlock()
+
+	if msg == nil {
+		select {
+		case <-w.Context.Done():
+			return 0, io.EOF
+		default:
+			w.mux.Lock()
+			defer w.mux.Unlock()
+			return 0, w.err
+		}
+	}
+
+	// TODO: don't truncate messages longer than the read buffer.
+	s := fmt.Sprintf("%v | %v", msg.PublishTime.Format(time.StampMilli), string(msg.Data))
+	return copy(p, []byte(s)), nil
+}
+
+func (w *pubsubTopicWatcher) Close() error {
+	// Stop receiving.
+	w.CancelFunc()
+	return w.Subscription.Delete(context.Background())
+}
+
+func (t *pubsubTopic) Stream(ctx context.Context) (io.ReadCloser, error) {
+	// Create subscription and wrap it in a ReadCloser to handle cleanup.
+	// TODO: make the subscription name prefix configurable.
+	return t.newPubsubTopicWatcher(ctx)
+}
+
+func (t *pubsubTopic) Write(ctx context.Context, _ int64, b []byte) error {
+	result := t.topic.Publish(ctx, &pubsub.Message{Data: b})
+	_, err := result.Get(ctx)
+	return err
+}
+
+func (t *pubsubTopic) Schema() *plugin.EntrySchema {
+	return plugin.NewEntrySchema(t, "topic").
+		SetMetaAttributeSchema(&pubsubTopicAttrMeta{}).
+		SetMetadataSchema(&pubsubTopicMetadata{}).
+		SetDescription(pubsubTopicDescription)
+}
+
+const pubsubTopicDescription = `
+A Cloud Pub/Sub topic. You can pipe text to it to publish messages.
```

Comment:
You could also refer to the `pubsub` dir description to avoid repeating yourself with the pipe example.